### PR TITLE
Add Dubbo Annotation Processor (Add deprecated method warn when called by user, #10850)

### PR DESCRIPTION
**Pull request by win120a**
## What is the purpose of the change
Trying to provide a solution of #10850.

## Brief changelog
Adds an annotation processor, which adds a line of code in every methods decorated by `@Deprecated`. Like getServiceName() in class URL of dubbo-common module (decompiled result): 
```java
@Deprecated
public String getServiceName() {
    DeprecatedMethodInvocationCounter.onDeprecatedMethodCalled("org.apache.dubbo.common.URL.getIntParameter(String key,int defaultValue)");
    return this.getServiceInterface();
}
```
Per suggestion from Albumen Kevin, the count of the deprecated method invocation is also recorded.

## Verifying this change
Use a decomplier (like Java Decompiler, or FernFlower decompiler in IntelliJ IDEA) to see the result.

### CODE REVIEW

**Comment by win120a** [comment]
@AlbumenJ PTAL.

**Review by AlbumenJ — CHANGES_REQUESTED** [review]
Please rebase your branch onto `3.2`

File: NOTICE [hunk 1]
```diff
@@ -4,11 +4,35 @@ Copyright 2018-2023 The Apache Software Foundation
...
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
```
Comment (AlbumenJ) [inline NOTICE:38]: No need to add notice if there is not NOTICE file in https://github.com/projectlombok/lombok.

File: dubbo-annotation-processor/src/main/java/org/apache/dubbo/annotation/handler/DeprecatedHandler.java [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+            // Only interested in methods.
+            if (!(element instanceof Symbol.MethodSymbol)) {
+                continue;
+            }
+
+            Symbol.ClassSymbol classSymbol = (Symbol.ClassSymbol) element.getEnclosingElement();
+
+            ASTUtils.addImportStatement(apContext, classSymbol, "org.apache.dubbo.common.logger", "LoggerFactory");
+            ASTUtils.addImportStatement(apContext, classSymbol, "org.apache.dubbo.common.logger", "ErrorTypeAwareLogger");
+            ASTUtils.addImportStatement(apContext, classSymbol, "org.apache.dubbo.common", "DeprecatedMethodInvocationCounter");
```
Comment (AlbumenJ) [inline dubbo-annotation-processor/src/main/java/org/apache/dubbo/annotation/handler/DeprecatedHandler.java:64]: Just insert `DeprecatedMethodCollector.collect(xxx)` here. Log and count in `DeprecatedMethodCollector.collect`.

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

**Review by AlbumenJ — COMMENTED** [review]
Can the inserted code also inserted into `xxx-source.jar`?

**Comment by win120a** [comment]
> Please rebase your branch onto `3.2` .

Rebased.

> Just insert DeprecatedMethodCollector.collect(xxx) here. Log and count in DeprecatedMethodCollector.collect.

Have moved all the logic to `org.apache.dubbo.common.DeprecatedMethodInvocationCounter#onDeprecatedMethodCalled()` .

> Can the inserted code also inserted into `xxx-source.jar`?

Unfortunately, it can't. It works as a compiler addon (an annotation processor). It directly modifes the Abstract Syntax Tree (i.e. the `.class` file), the source code remains untouched. Since the source plugin in Maven just copies the source, the generated sources.jar will not appear the modified code.

**Review by AlbumenJ — COMMENTED** [review]
@chickenlj PTAL

**Comment by win120a** [comment]
@AlbumenJ PTAL.
The scope configuration of pom.xml is required to import javac package (tools.jar), which causes 'bugs' in SonarLint.

File: dubbo-annotation-processor/src/main/java/org/apache/dubbo/annotation/permit/Permit.java [hunk 3]
```diff
@@ -0,0 +1,446 @@
...
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
```
Comment (AlbumenJ) [inline dubbo-annotation-processor/src/main/java/org/apache/dubbo/annotation/permit/Permit.java:42]: Leave the origin license here and add exclude in license checker.

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

**Review by AlbumenJ — APPROVED** [review]
LGTM

**Pull request merged**